src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.cs

```diff
@@ -1047,6 +1047,66 @@ internal void OnAdditionalDocumentTextUpdatedOnDisk(DocumentId documentId)
             return this.ServiceProvider.GetService(typeof(TService)) as TInterface;
         }
 
+        internal override bool CanAddProjectReference(ProjectId referencingProject, ProjectId referencedProject)
+        {
+            _foregroundObject.AssertIsForeground();
+
+            IVsHierarchy referencingHierarchy;
+            IVsHierarchy referencedHierarchy;
+            if (!TryGetHierarchy(referencingProject, out referencingHierarchy) ||
+                !TryGetHierarchy(referencedProject, out referencedHierarchy))
+            {
+                return false;
+            }
+
+            var referencingProjectFlavor3 = referencingHierarchy as IVsProjectFlavorReferences3;
+            var referencedProjectFlavor3 = referencedHierarchy as IVsProjectFlavorReferences3;
+
+            if (referencingProjectFlavor3 != null && referencedProjectFlavor3 != null)
```

Comment:
This doesn't look right. If the referencing project (say portable) implements IVsProjectFlavorReferences3, but the referenced (say C++) project doesn't - you're basically skipping all portable checks here and falling back to the default P2P matrix. And vice versa.